### ISSUE TITLE
`sort-by` gives scary error message on bad first argument

### ISSUE THREAD

**Issue by TheBizzle** [NetLogo/NetLogo#1389]
I got this error while running a model:

```
error (MatchError)
 while observer running FOREACH
  called by procedure GO
  called by Button 'go'

NetLogo is unable to supply you with more details about this error.  Please report the problem 
 at https://github.com/NetLogo/NetLogo/issues, or to bugs@ccl.northwestern.edu, and paste the 
contents of this window into your report. 

scala.MatchError: - (of class java.lang.Character)
 at org.nlogo.nvm.AnonymousProcedure.$anonfun$checkAgentClass$1(AnonymousProcedure.scala:33)
 at org.nlogo.nvm.AnonymousProcedure.$anonfun$checkAgentClass$1$adapted(AnonymousProcedure.scala:33)
 at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
 at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
 at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
 at scala.collection.immutable.StringOps.foreach(StringOps.scala:29)
 at scala.collection.TraversableLike.map(TraversableLike.scala:234)
 at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
 at scala.collection.immutable.StringOps.map(StringOps.scala:29)
 at org.nlogo.nvm.AnonymousProcedure.checkAgentClass(AnonymousProcedure.scala:33)
 at org.nlogo.nvm.AnonymousProcedure.checkAgentClass$(AnonymousProcedure.scala:26)
 at org.nlogo.nvm.AnonymousReporter.checkAgentClass(AnonymousProcedure.scala:74)
 at org.nlogo.nvm.AnonymousReporter.report(AnonymousProcedure.scala:96)
 at org.nlogo.nvm.AnonymousReporter.report(AnonymousProcedure.scala:93)
 at org.nlogo.prim.etc._sortby$MyComparator.compare(_sortby.scala:53)
 at java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
 at java.util.TimSort.sort(TimSort.java:234)
 at java.util.Arrays.sort(Arrays.java:1512)
 at java.util.ArrayList.sort(ArrayList.java:1454)
 at java.util.Collections.sort(Collections.java:175)
 at org.nlogo.prim.etc._sortby.report(_sortby.scala:37)
 at org.nlogo.nvm.Instruction.argEvalList(Instruction.scala:151)
 at org.nlogo.prim.etc._foreach.$anonfun$perform$1(_foreach.scala:16)
 at org.nlogo.prim.etc._foreach.$anonfun$perform$1$adapted(_foreach.scala:15)
 at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
 at scala.collection.immutable.Range.foreach(Range.scala:156)
 at scala.collection.TraversableLike.map(TraversableLike.scala:234)
 at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
 at scala.collection.AbstractTraversable.map(Traversable.scala:104)
 at org.nlogo.prim.etc._foreach.perform(_foreach.scala:15)
 at org.nlogo.nvm.Context.stepConcurrent(Context.java:107)
 at org.nlogo.nvm.ConcurrentJob.step(ConcurrentJob.scala:65)
 at org.nlogo.job.JobThread.runPrimaryJobs(JobThread.scala:133)
 at org.nlogo.job.JobThread.$anonfun$run$1(JobThread.scala:68)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
 at scala.util.control.Exception$Catch.apply(Exception.scala:224)
 at org.nlogo.api.Exceptions$.handling(Exceptions.scala:41)
 at org.nlogo.job.JobThread.run(JobThread.scala:66)

NetLogo 6.0.1
main: org.nlogo.app.AppFrame
thread: JobThread
Java HotSpot(TM) 64-Bit Server VM 1.8.0_121 (Oracle Corporation; 1.8.0_121-b13)
operating system: Linux 4.4.0-21-generic (amd64 processor)
Scala version 2.12.1
JOGL: (3D View not initialized)
OpenGL Graphics: (3D View not initialized)
model: Untitled

12:17:26.888 SwitchedTabsEvent (org.nlogo.app.Tabs) AWT-EventQueue-0
12:17:26.885 RuntimeErrorEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:26.755 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:26.658 AddJobEvent (org.nlogo.window.ButtonWidget) AWT-EventQueue-0
12:17:26.625 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:26.485 InputBoxLoseFocusEvent (org.nlogo.window.ButtonWidget) AWT-EventQueue-0
12:17:26.425 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:26.225 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:26.024 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:17:25.824 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
```

A simple way to reproduce this is to create some turtles and then run the code `sort-by [who] turtles` from the Command Center.  It's right that we should get an error from doing that, since the function given to `sort-by` is supposed to be a binary function, but this isn't the error we should be giving.

**Comment by mrerrormessage** [comment]
Thanks for reporting, this definitely needs to be fixed!

**Comment by mrerrormessage** [comment]
OK, I've fixed the exception in #1394. `sort-by [who] turtles` still throws a runtime error on desktop NetLogo - specifically "this code can't be run by the observer, only a turtle", since `sort-by` doesn't set agentset context. Does NLW work differently? The assumption here is that the first input to `sort-by` is an anonymous reporter, not a reporter block, and therefore it will not have context set.

**Comment by TheBizzle** [comment]
Yeah, that's right.  This was really about fixing the error message.  Rather than a stacktrace and a very unclear error message, we should be getting told something along the lines of "this code can't be run by the observer, only a turtle" in these cases.

So :+1:

**Comment by mrerrormessage** [comment]
Fixed in #1394
